util/Pos/Util/Timer.hs

```diff
@@ -22,38 +22,45 @@ module Pos.Util.Timer
     , setTimerDuration
     , waitTimer
     , startTimer
+    , asDynamicTimer
     ) where
 
+import           Data.Time.Units (Microsecond)
 import           Control.Concurrent (modifyMVar_)
 import           Control.Concurrent.STM (readTVar, registerDelay, retry)
+import           Pos.Util.DynamicTimer(DynamicTimer(..))
 import           Universum
 
 data Timer = Timer
-  { timerDuration   :: !(MVar Int)
-  , timerSempaphore :: !(TVar (TVar Bool))
+  { timerDuration   :: !(MVar Microsecond)
+  , timerSemaphore :: !(TVar (TVar Bool))
   }
 
 -- | Create a new, inactive Timer given the number of microseconds. In order to
 -- activate it 'startTimer' needs to be called.
-newTimer :: MonadIO m => Int -> m Timer
+newTimer :: MonadIO m => Microsecond -> m Timer
 newTimer n = Timer <$> newMVar n <*> (newTVarIO True >>= newTVarIO)
 
 -- | Set the duration of a timer to a specified number of microseconds. Note
 -- that if the timer is already started, calling this function wont't restart
 -- it.
-setTimerDuration :: MonadIO m => Timer -> Int -> m ()
+setTimerDuration :: MonadIO m => Timer -> Microsecond -> m ()
 setTimerDuration Timer{..} n = liftIO $ modifyMVar_ timerDuration $ \_ -> return $! n
 
 -- | Wait for the duration associated with the Timer that passed since the last
 -- time 'startTimer' was called.
 waitTimer :: Timer -> STM ()
 waitTimer Timer{..} = do
-  done <- readTVar =<< readTVar timerSempaphore
+  done <- readTVar =<< readTVar timerSemaphore
   unless done retry
 
 -- Start the timer. If the function is called before @t + n@, where @t@ is the
 -- time startTimer was called for the last time and @n@ is the number of
 -- microseconds associated with the Timer, the timer is restarted.
 startTimer :: MonadIO m => Timer -> m ()
-startTimer Timer{..} = liftIO (registerDelay =<< readMVar timerDuration)
-    >>= atomically . writeTVar timerSempaphore
+startTimer Timer{..} = liftIO (registerDelay . fromIntegral =<< readMVar timerDuration)
+    >>= atomically . writeTVar timerSemaphore
```

Comment:
I think the comment is wrong. `waitTimer` will take the current delay `TVar` from the outer `TVar`, but `startTimer` will replace the contents of the outer `TVar`, so the earlier `waitTimer` will still only wait for the original duration. What do you think? Am I missing something?

Also, would be good to really consolidate this with `DynamicTimer`. I think we only need `DynamicTimer`.

Reply:
That's true. `threadDelay` (which is used) does not expose any interface to manipulate the timeout.  

I agree that we actually only need the `DynamicTimer`.  This will make the other timers where slot duration is used more elastic on slot duration changes.